### PR TITLE
optimally rotated RMSD + grads

### PR DESCRIPTION
**Pull request by proteneer**
@mpharrigan did the initial tensorflow version. I just did some copy-pasta into jax.

This PR:
- [x] Implements the optimally rotated RMSD code
- [x] Compute derivatives of the opt_rot_rmsd w.r.t. geometries
- [x] Tests the above code

### CODE REVIEW

File: tests/test_rmsd.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        opt_init, opt_update, get_params = optimizers.sgd(5e-2)
+
+        x0 = onp.array([
+            [1.0, 0.2, 3.3], # H
+            [-0.6,-1.1,-0.9],# C
+            [3.4, 5.5, 0.2], # H
+            [3.6, 5.6, 0.6], # H
+        ], dtype=onp.float64)
+
+        x1 = onp.array([
```
Comment (bp-kelley) [inline tests/test_rmsd.py]: Might be clearer as x0 + onp.random...

File: timemachine/observables/rmsd.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import jax.scipy
+import jax.numpy as np
+
+
+def optimal_rotational_quaternion(r):
+    """Just need the largest eigenvalue of this to minimize RMSD over rotations
+    
+    References
+    ----------
+    [1] http://dx.doi.org/10.1002/jcc.20110
```
Comment (bp-kelley) [inline timemachine/observables/rmsd.py]: It’s amazing to me how this gets recycled, although the gradient is a nice touch.

This is the earliest ref I have for these methods:

http://www.ccl.net/cca/software/SOURCES/C/quaternion-mol-fit/quatfit.txt.shtml

File: tests/test_rmsd.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            [1.0, 0.2, 3.3], # H
+            [-0.6,-1.1,-0.9],# C
+            [3.4, 5.5, 0.2], # H
+            [3.6, 5.6, 0.6], # H
+        ], dtype=onp.float64) + onp.random.rand(x0.shape[0],3)*10
+
+        grad_fn = jax.jit(jax.grad(rmsd.opt_rot_rmsd, argnums=(0,)))
+        opt_state = opt_init(x0)
+
+        for i in range(1500):
```
Comment (bp-kelley) [inline tests/test_rmsd.py]: Is this just for convergence?  This seems to be a lot of iterations.

File: tests/test_rmsd.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            g = grad_fn(get_params(opt_state), x1)[0]
+            opt_state = opt_update(i, g, opt_state)
+
+        x_final = get_params(opt_state)
+
+        assert rmsd.opt_rot_rmsd(x_final, x1) < 0.1
```
Comment (bp-kelley) [inline tests/test_rmsd.py]: .1 seems high to me.

File: tests/test_rmsd.py (hunk 1 above)
Comment (proteneer) [reply tests/test_rmsd.py]: I can fix this in a patch

File: tests/test_rmsd.py (hunk 3 above)
Comment (proteneer) [reply tests/test_rmsd.py]: I'm not sure how much is necessary, but 1500 iterations take less than 2 seconds to run when JITted. So it's very fast. I just wanted to set a high enough number to avoid convergence issues. With quasi-newton methods we can do a much smaller number of iterations, but this is just gradient descent here.

File: tests/test_rmsd.py (hunk 4 above)
Comment (proteneer) [reply tests/test_rmsd.py]: I thought about this for a while too. The problem is that I can't prove for sure that the optimally rotated RMSD is actually a convex function (vanilla RMSD is). I tested this for about 50 times and it varies between 0.005 to 0.02. So I don't think it's strictly convex at least in the numerical sense (else it would converge exactly to 1e-7 under double precision)

File: timemachine/observables/rmsd.py (hunk 2 above)
Comment (proteneer) [reply timemachine/observables/rmsd.py]: Everyday i'm standin' on the shoulders of giants

**Comment by proteneer** [comment]
@bp-kelley I think I merged this too quickly.

File: timemachine/observables/rmsd.py (hunk 2 above)
Comment (bp-kelley) [reply timemachine/observables/rmsd.py]: This was my goto for rmsd:  https://theobald.brandeis.edu/qcp/

Note that in the later papers one of the coauthors is now head of nx at novartis...

https://theobald.brandeis.edu/qcp/

Comment (proteneer) [reply timemachine/observables/rmsd.py]: Nice! I ported that to the GPU back in 2012 during grad school:

https://onlinelibrary.wiley.com/doi/10.1002/jcc.23110
